Clients/Win32 Player/Code/Game/Audio/SongCache.cpp

```
@@ -122,3 +190,92 @@ SongResourceInfo::~SongResourceInfo()
 {
     AudioSystem::instance->ReleaseRawSong(m_songData);
 }
+
+//-----------------------------------------------------------------------------------
+SongID SongCache::FindLeastAccessedSong()
+{
+    double lowestAccessTime = -1;
+    SongID leastAccessedSong = 0;
+
+    //Try to find a song that is loaded in and has been played. Otherwise find one that has been loaded
+    for (std::map<SongID, SongResourceInfo>::iterator i = m_songCache.begin(); i != m_songCache.end(); ++i)
+    {
+        SongResourceInfo& info = i->second;
+        //Has this song been loaded yet? Has it been loaded and not played? Is the last load time less than what we have?
+        if (((info.m_timeLastAccessedMS < lowestAccessTime) || lowestAccessTime == -1) && info.m_timeLastAccessedMS != -1 && info.m_songData && info.m_status != PLAYING)
+        {
+            lowestAccessTime = info.m_timeLastAccessedMS;
+            leastAccessedSong = info.m_songID;
+        }
+        else if (info.m_songData && info.m_status != PLAYING && leastAccessedSong == -1)
+        {
+            leastAccessedSong = info.m_songID;
+        }
+    }
+
+    return leastAccessedSong;
+}
+
+//-----------------------------------------------------------------------------------
+void SongCache::RemoveFromCache(const SongID songID)
+{
+    if (songID == 0)
+    {
+        return;
```

If we ever get an invalid result to remove from the cache, silently fail? How does this end up working out from an end user's perspective?